packages/react-dom/src/__tests__/ReactMount-test.js

```diff
@@ -799,6 +799,62 @@ describe('ReactMount', () => {
     );
   });
 
+  it('should warn with special characters in the JSX string output', () => {
+    const div = document.createElement('div');
+    const markup = ReactDOMServer.renderToString(
+      <div>{'SSRMismatchTest special characters: \'"\t\n'}</div>,
+    );
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <span />
+          {'SSRMismatchTest special characters: \'"\t\n'}
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      // TODO: Currently, the special characters in the JSX string output are not escaped, the output looks invalid.
+      'Warning: Expected server HTML to contain a matching <span> in <div>.\n\n' +
+        '  <div data-reactroot="">\n' +
+        "-   {'SSRMismatchTest special characters: '\"\t\n'}\n" +
```

Comment:
Yeah, this looks weird.

What does Jest do when you diff two strings? e.g. `expect(stringWithSpecialChars).toEqual(stringWithoutSpecialChars)`?

They likely already solved this somehow. We could use whatever solution they came up with.

Reply:
I'm not talking about Jest, I'm talking about showing a warning with invalid syntax to the developers, e.g.
```
<div data-attr=""many""""unescaped""quotes""""><<<<<<looks like a tag but it's not, it's text content>>>>></div>
```
Maybe it's an edge case, but I'm trying to think about these as well.

Reply:
Oh sorry you mean how Jest escapes the values — these special chars like tab and newline were printed verbatim, unescaped.

Reply:
But the special chars within "expected" and "received" strings are, contrary to our case, not printed in a JavaScript/JSX or HTML context (like within quotes or tags), they are printed on separate lines without any delimiters. See screenshots below where I added `FOOBAR` to make the tests fail:
<img width="1541" alt="screen shot 2018-06-19 at 7 42 55 am" src="https://user-images.githubusercontent.com/498274/41604656-9c2165a4-7394-11e8-8d22-aa62991bd3a7.png">
<img width="1541" alt="screen shot 2018-06-19 at 7 43 07 am" src="https://user-images.githubusercontent.com/498274/41604669-a117c97c-7394-11e8-85f5-3a690465bcf9.png">

Reply:
Here's what [Jest does in their `jest-matcher-utils` package](https://github.com/facebook/jest/blob/7542f24f421ab9694eb3bd5c0a67ff79ad223cb4/packages/jest-matcher-utils/src/index.js#L59-L89):
```js
export const printReceived = (object: any) =>
  RECEIVED_COLOR(highlightTrailingWhitespace(stringify(object)));
export const printExpected = (value: any) =>
  EXPECTED_COLOR(highlightTrailingWhitespace(stringify(value)));
```
Within `stringify` they use `pretty-format` library with plugins for DOM and React element printing – this library goes extra mile to make the output beautiful – including [using](https://github.com/facebook/jest/blob/7542f24f421ab9694eb3bd5c0a67ff79ad223cb4/packages/pretty-format/src/plugins/lib/markup.js#L79-L82) their own [`escapeHTML`](https://github.com/facebook/jest/blob/7542f24f421ab9694eb3bd5c0a67ff79ad223cb4/packages/pretty-format/src/plugins/lib/escape_html.js) function, as well as many functions to print HTML-like tags.

The more I look at this `pretty-format/src/plugins/lib/markup.js` file, the more it resembles what I wrote here by hand, but without the diff part (the plus and minus prefixes on certain lines). I'd love to reuse something from there, but this will require `react` to depend on `pretty-format` which I don't think would be a good idea.

By the way, they [use the unicode ellipsis `…` in `printElementAsLeaf`](https://github.com/facebook/jest/blob/7542f24f421ab9694eb3bd5c0a67ff79ad223cb4/packages/pretty-format/src/plugins/lib/markup.js#L132-L144).